src/frequenz/sdk/timeseries/battery_pool/_methods.py

```diff
@@ -0,0 +1,265 @@
+
+    _result_channel: Broadcast[T]
+    _cached_metrics: Dict[int, ComponentMetricsData]
+
+    _update_task: asyncio.Task[None]
+    _send_task: asyncio.Task[None]
+
+    def __init__(
+        self,
+        working_batteries: Set[int],
+        metric_calculator: MetricCalculator[T],
+        min_update_interval: timedelta,
+    ) -> None:
+        """Create class instance.
+
+        Args:
+            working_batteries: Set of the working batteries
+            metric_calculator: Module that tells how to aggregate the component metrics.
+            min_update_interval: Minimum frequency for sending update about the change.
+        """
+        self._metric_calculator = metric_calculator
+        self._working_batteries = working_batteries.intersection(
+            metric_calculator.batteries
+        )
+        self._result_channel = Broadcast[T](
+            name=SendOnUpdate.name() + "_" + metric_calculator.name(),
+            resend_latest=True,
+        )
+
+        self._update_event = asyncio.Event()
+        self._cached_metrics = {}
+
+        self._update_task = asyncio.create_task(self._update_and_notify())
+        self._send_task = asyncio.create_task(self._send_on_update(min_update_interval))
+        self._pending_data_fetchers: set[
+            asyncio.Task[ComponentMetricsData | None]
+        ] = set()
+
+    @classmethod
+    def name(cls) -> str:
+        """Get name of the method.
+
+        Returns:
+            Name of the method.
+        """
+        return "SendOnUpdate"
+
+    def new_receiver(self, maxsize: Optional[int]) -> Receiver[T]:
+        """Return new receiver for the aggregated metric results.
+
+        Args:
+            maxsize: Buffer size of the receiver
+
+        Returns:
+            Receiver for the metric results.
+        """
+        if maxsize is None:
+            return self._result_channel.new_receiver()
+        return self._result_channel.new_receiver(maxsize=maxsize)
+
+    def update_working_batteries(self, new_working_batteries: Set[int]):
+        """Update set of the working batteries.
+
+        Recalculate metric if set changed.
+
+        Args:
+            new_working_batteries: Set of the working batteries.
+        """
+        # MetricCalculator can discard some batteries. (for example because they
+        # don't have adjacent inverter)
+        new_set = new_working_batteries.intersection(self._metric_calculator.batteries)
+
+        if new_set != self._working_batteries:
+            self._working_batteries = new_set
+            self._update_event.set()
+
+    async def stop(self) -> None:
+        """Stop the all pending async task."""
+        await asyncio.gather(
+            *[cancel_and_await(task) for task in self._pending_data_fetchers],
+            return_exceptions=True,
+        )
+        await asyncio.gather(
+            *[cancel_and_await(self._send_task), cancel_and_await(self._update_task)],
+            return_exceptions=True,
+        )
```

Comment:
So you are just silently swallowing exceptions? Maybe at least print a log if there are any?

In any case all this should be handled by the `BackgroundService` class (if I ever find some time to take care of it :grimacing: ).

Reply:
Actually it should not raise exception, because exception is catched in the `cancel_and_await`. 
If for some reason it will raise and exception then it is something really really wrong and should not be catched.
So I removed `return_exceptions=True`, thanks